File: backend/app/api/routes/tasks.py
"""Tasks API endpoints."""

from typing import List, Optional
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.task import (
    TaskCreate,
    TaskPriorityUpdate,
    TaskResponse,
    TaskStatusUpdate,
    TaskUpdate,
)
from app.services.task_service import TaskService

router = APIRouter(prefix="/tasks", tags=["Tasks"])


@router.get(
    "",
    response_model=List[TaskResponse],
    summary="List tasks",
    description="Retrieves tasks belonging to the current user's projects with filtering by project, status, priority, and search.",
)
def list_tasks(
    project_id: Optional[str] = Query(None, description="Filter by project ID"),
    status: Optional[str] = Query(None, description="Filter by status ('todo', 'in_progress', 'done')"),
    priority: Optional[str] = Query(None, description="Filter by priority ('low', 'medium', 'high', 'critical')"),
    search: Optional[str] = Query(None, description="Search term in title, description, or project"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> List[TaskResponse]:
    return TaskService.list_tasks(
        db,
        user_id=current_user.id,
        project_id=project_id,
        status_filter=status,
        priority_filter=priority,
        search=search,
    )


@router.post(
    "",
    response_model=TaskResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new task",
    description="Creates a task under a specified project after verifying ownership.",
)
def create_task(
    data: TaskCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> TaskResponse:
    return TaskService.create_task(db, user=current_user, data=data)


@router.post(
    "/batch",
    response_model=List[TaskResponse],
    status_code=status.HTTP_201_CREATED,
    summary="Batch create tasks (e.g. approved AI generated tasks)",
    description="Creates multiple tasks at once under a project.",
)
def create_tasks_batch(
    tasks_data: List[TaskCreate],
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> List[TaskResponse]:
    created = []
    for item in tasks_data:
        task_res = TaskService.create_task(db, user=current_user, data=item)
        created.append(task_res)
    return created


@router.get(
    "/{task_id}",
    response_model=TaskResponse,
    summary="Get task by ID",
    description="Retrieves a single task, verifying project ownership or assignee rights.",
)
def get_task(
    task_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> TaskResponse:
    return TaskService.get_task(db, task_id=task_id, user_id=current_user.id)


@router.put(
    "/{task_id}",
    response_model=TaskResponse,
    summary="Update task",
    description="Updates task properties with authorization verification.",
)
def update_task(
    task_id: str,
    data: TaskUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> TaskResponse:
    return TaskService.update_task(db, task_id=task_id, user=current_user, data=data)


@router.patch(
    "/{task_id}/status",
    response_model=TaskResponse,
    summary="Quick update task status",
    description="Updates only the workflow status of a task ('todo', 'in_progress', 'done').",
)
def update_task_status(
    task_id: str,
    status_data: TaskStatusUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> TaskResponse:
    return TaskService.update_task(
        db, task_id=task_id, user=current_user, data=TaskUpdate(status=status_data.status)
    )


@router.patch(
    "/{task_id}/priority",
    response_model=TaskResponse,
    summary="Quick update task priority",
    description="Updates only the priority of a task.",
)
def update_task_priority(
    task_id: str,
    priority_data: TaskPriorityUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> TaskResponse:
    return TaskService.update_task(
        db, task_id=task_id, user=current_user, data=TaskUpdate(priority=priority_data.priority)
    )


@router.delete(
    "/{task_id}",
    summary="Delete task",
    description="Removes a task from a project.",
)
def delete_task(
    task_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    return TaskService.delete_task(db, task_id=task_id, user=current_user)

File: frontend/src/services/tasks.ts
import { apiClient } from './api';
import { Task, TaskPriority, TaskStatus } from '../types';

export interface CreateTaskPayload {
  title: string;
  description?: string;
  project_id: string;
  assignee_id?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  due_date?: string;
  tags?: string;
  estimated_hours?: number;
}

export interface UpdateTaskPayload {
  title?: string;
  description?: string;
  project_id?: string;
  assignee_id?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  due_date?: string;
  tags?: string;
  estimated_hours?: number;
}

function mapBackendTask(t: any): Task {
  return {
    id: t.id,
    projectId: t.project_id,
    projectName: t.project_name,
    title: t.title,
    description: t.description || '',
    status: t.status,
    priority: t.priority,
    dueDate: t.due_date,
    assigneeId: t.assignee_id,
    assigneeName: t.assignee_name,
    tags: t.tags || 'task',
    estimatedHours: t.estimated_hours,
    createdAt: t.created_at,
    updatedAt: t.updated_at,
  };
}

export const taskService = {
  async listTasks(params?: {
    projectId?: string;
    status?: string;
    priority?: string;
    search?: string;
  }): Promise<Task[]> {
    const query = new URLSearchParams();
    if (params?.projectId && params.projectId !== 'all') query.append('project_id', params.projectId);
    if (params?.status && params.status !== 'all') query.append('status', params.status);
    if (params?.priority && params.priority !== 'all') query.append('priority', params.priority);
    if (params?.search) query.append('search', params.search);

    const queryString = query.toString() ? `?${query.toString()}` : '';
    const raw = await apiClient<any[]>(`/tasks${queryString}`);
    return raw.map(mapBackendTask);
  },

  async getTask(id: string): Promise<Task> {
    const raw = await apiClient<any>(`/tasks/${id}`);
    return mapBackendTask(raw);
  },

  async createTask(data: CreateTaskPayload): Promise<Task> {
    const raw = await apiClient<any>('/tasks', {
      method: 'POST',
      body: JSON.stringify(data),
    });
    return mapBackendTask(raw);
  },

  async createTasksBatch(tasks: CreateTaskPayload[]): Promise<Task[]> {
    const raw = await apiClient<any[]>('/tasks/batch', {
      method: 'POST',
      body: JSON.stringify(tasks),
    });
    return raw.map(mapBackendTask);
  },

  async updateTask(id: string, data: UpdateTaskPayload): Promise<Task> {
    const raw = await apiClient<any>(`/tasks/${id}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
    return mapBackendTask(raw);
  },

  async updateTaskStatus(id: string, status: TaskStatus): Promise<Task> {
    const raw = await apiClient<any>(`/tasks/${id}/status`, {
      method: 'PATCH',
      body: JSON.stringify({ status }),
    });
    return mapBackendTask(raw);
  },

  async updateTaskPriority(id: string, priority: TaskPriority): Promise<Task> {
    const raw = await apiClient<any>(`/tasks/${id}/priority`, {
      method: 'PATCH',
      body: JSON.stringify({ priority }),
    });
    return mapBackendTask(raw);
  },

  async deleteTask(id: string): Promise<{ message: string }> {
    return apiClient<{ message: string }>(`/tasks/${id}`, {
      method: 'DELETE',
    });
  },
};
